Framework/Crystal/src/OptimizeLatticeForCellType.cpp

```diff
@@ -283,153 +222,127 @@ void OptimizeLatticeForCellType::exec() {
 }
 //-----------------------------------------------------------------------------------------
 /**
-  @param  inname       Name of Filename containing peaks
   @param  cell_type    cell type to optimize
-  @param  params       optimized cell parameters
-  @return  chisq of optimization
+  @param  lat          optimized cell parameters
+  @return  fun_str     string of parameters for fitting
 */
-double OptimizeLatticeForCellType::optLatticeSum(std::string inname,
-                                                 std::string cell_type,
-                                                 std::vector<double> &params) {
+std::string OptimizeLatticeForCellType::inParams(std::string cell_type,
+                                                 std::vector<double> &lat) {
+  std::ostringstream fun_str;
+  // TODO remove next 3 lines when PointGroup is changed
+  if (cell_type == "Rhombohedral")
+    fun_str << "name=LatticeFunction,CrystalSystem=Trigonal";
+  else
+    fun_str << "name=LatticeFunction,CrystalSystem=" << cell_type;
+
   std::vector<double> lattice_parameters;
   lattice_parameters.assign(6, 0);
   if (cell_type == ReducedCell::CUBIC()) {
-    lattice_parameters[0] = params[0];
-    lattice_parameters[1] = params[0];
-    lattice_parameters[2] = params[0];
+    fun_str << ",a=" << lat[0];
+  } else if (cell_type == ReducedCell::TETRAGONAL()) {
+    fun_str << ",a=" << lat[0];
+    fun_str << ",c=" << lat[2];
+  } else if (cell_type == ReducedCell::ORTHORHOMBIC()) {
+    fun_str << ",a=" << lat[0];
+    fun_str << ",b=" << lat[1];
+    fun_str << ",c=" << lat[2];
+  } else if (cell_type == ReducedCell::RHOMBOHEDRAL()) {
+    fun_str << ",a=" << lat[0];
+    fun_str << ",Alpha=" << lat[3];
+  } else if (cell_type == ReducedCell::HEXAGONAL()) {
+    fun_str << ",a=" << lat[0];
+    fun_str << ",c=" << lat[2];
+  } else if (cell_type == ReducedCell::MONOCLINIC()) {
+    fun_str << ",a=" << lat[0];
+    fun_str << ",b=" << lat[1];
+    fun_str << ",c=" << lat[2];
+    fun_str << ",Beta=" << lat[4];
+  } else if (cell_type == ReducedCell::TRICLINIC()) {
+    fun_str << ",a=" << lat[0];
+    fun_str << ",b=" << lat[1];
+    fun_str << ",c=" << lat[2];
+    fun_str << ",Alpha=" << lat[3];
+    fun_str << ",Beta=" << lat[4];
+    fun_str << ",Gamma=" << lat[5];
+  }
```

Comment:
I think it's possible to simplify this a bit more, even getting rid of the `if...else`-construct, because this is already implemented in `LatticeFunction` itself. The method I have in mind would probably look like this:

```
API::IFunction_sptr getLatticeFunction(const std::string &cellType, const UnitCell &cell) const {
  std::ostringstream fun_str;

  // Construct the string like it's done now, using 'name=LatticeFunction,CrystalSystem=...

  API::IFunction_sptr rawFunction = API::FunctionFactory.Instance().createInitialized(fun_str.str())

  API::ILatticeFunction_sptr latticeFunction = boost::dynamic_pointer_cast<API::ILatticeFunction>(rawFunction);
  if(latticeFunction) {
     latticeFunction->setUnitCell(cell);
  }

  return latticeFunction;
}
```

In the calling code, the method would be called like this:

```
API::IFunction_sptr latticeFunction = getLatticeFunction(cell_type, peakWS->sample().getOrientedLattice());
...
fit_alg->setProperty("Function", latticeFunction);
...
```

I haven't tried to compile this, but it should work and remove more code from the class. I'll have to check whether something similar is possible for processing the output, but it's been a while since I implemented this and I'm not sure the errors are handled.